x-pack/plugins/lens/public/indexpattern_datasource/dimension_panel/truncated_label.tsx

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import React from 'react';
+import { EuiMark } from '@elastic/eui';
+import { EuiHighlight } from '@elastic/eui';
+
+const createContext = () =>
+  document.createElement('canvas').getContext('2d') as CanvasRenderingContext2D;
+
+// extracted from getTextWidth for performance
+const context = createContext();
+
+const getTextWidth = (text: string, font: string) => {
+  const ctx = context ?? createContext();
+  ctx.font = font;
+  const metrics = ctx.measureText(text);
+  return metrics.width;
+};
+
+const truncateLabel = (
+  width: number,
+  font: string,
+  label: string,
+  approximateLength: number,
+  labelFn: (label: string, length: number) => string
+) => {
+  let output = labelFn(label, approximateLength);
+  while (getTextWidth(output, font) > width) {
+    approximateLength = approximateLength - 1;
+    output = labelFn(label, approximateLength);
+  }
+  return output;
+};
+
+export const TruncatedLabel = ({
+  label,
+  width,
+  search,
+  font,
+}: {
+  label: string;
+  search: string;
+  width: number;
+  font: string;
+}) => {
+  const textWidth = getTextWidth(label, font);
```

Comment:
This is calculating the text size on every render, seems like a good thing to memoize. The whole component could be memoized, but also we should memoize on the `label` and `font` keys internally. So as the user filters the label it should not recalculate the width each keystroke.